swift/aperture/util.swift

```diff
@@ -1,26 +1,213 @@
 import AppKit
 import AVFoundation
 
-private final class StandardErrorOutputStream: TextOutputStream {
-  func write(_ string: String) {
-    FileHandle.standardError.write(string.data(using: .utf8)!)
+
+// MARK: - SignalHandler
+struct SignalHandler {
+  struct Signal: Hashable {
+    static let hangup = Signal(rawValue: SIGHUP)
+    static let interrupt = Signal(rawValue: SIGINT)
+    static let quit = Signal(rawValue: SIGQUIT)
+    static let abort = Signal(rawValue: SIGABRT)
+    static let kill = Signal(rawValue: SIGKILL)
+    static let alarm = Signal(rawValue: SIGALRM)
+    static let termination = Signal(rawValue: SIGTERM)
+    static let userDefined1 = Signal(rawValue: SIGUSR1)
+    static let userDefined2 = Signal(rawValue: SIGUSR2)
```

Comment:
I initially had this as an `enum`, but to be able to create a new enum instance, I need to manually implement an `init()` that switches over all the values and does the inverse of `rawValue`. I didn't think it was worth all the bloat. The only benefit of an `enum` here would be exhaustiveness checking when switching over signals, right? (@LarsJK @LinusU Thoughts?) In which case, I don't see it being very useful as I can't imagine anyone would want to switch over all the signals.


```swift
enum Signal: Hashable {
  case hangup
  case interrupt
  case quit
  case abort
  case kill
  case alarm
  case termination
  case userDefined1
  case userDefined2
  case custom(Int)

  /// Signals that cause the process to exit
  static let exitSignals = [
    hangup,
    interrupt,
    quit,
    abort,
    alarm,
    termination
  ]

  var rawValue: Int32 {
    switch self {
    case .hangup:
      return SIGHUP
    case .interrupt:
      return SIGINT
    case .quit:
      return SIGQUIT
    case .abort:
      return SIGABRT
    case .kill:
      return SIGKILL
    case .alarm:
      return SIGALRM
    case .termination:
      return SIGTERM
    case .userDefined1:
      return SIGUSR1
    case .userDefined2:
      return SIGUSR2
    case .custom(let signal):
      return Int32(signal)
    }
  }
}
```

Reply:
I don't see any clear benefits of using an enum either, but note that you could get RawRepresentable stuff for free:

```swift
enum Signal: Int32 {
  case hangup = 1 // SIGHUP
  case interrupt = 2 // SIGINT
  case quit = 3 // SIGQUIT
  case abort = 6 // SIGABRT
  case kill = 9 // SIGKILL
  case alarm = 14 // SIGALRM
  case termination = 15 // SIGTERM
  case userDefined1 = 30 // SIGUSR1
  case userDefined2 = 31 // SIGUSR2
}
```

Reply:
@LarsJK I considered that but weren't sure whether hard-coding the signal numbers are safe (?). According to Linux manpages `SIGUSR1` can be any of the following values: `SIGUSR1   30,10,16`, but maybe they're stable on macOS? If this is safe, I think it's much more elegant.

I do wish `enum` supported variables as raw values (as long as they resolve to the correct type, of course)... Do you think it's worth doing a Swift evolution proposal about that?

Reply:
You are right this may not be safe.. Struct is a better approach.

As long as the variables are read only I think it would be nice to be able to do something like this 👍